eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/Imp/UtxosSpec.hs

```diff
@@ -0,0 +1,157 @@
+    (inject $ Coin 1_000_000)
+    & dataHashTxOutL .~ SJust (hashData @era . Data $ P1.I 3)
+
+mkRefTxOut :: BabbageEraTxOut era => ImpTestM era (TxOut era)
+mkRefTxOut = do
+  kpPayment <- lookupKeyPair =<< freshKeyHash
+  kpStaking <- lookupKeyPair =<< freshKeyHash
+  pure $ mkBasicTxOut (mkAddr (kpPayment, kpStaking)) (inject $ Coin 100)
+    & referenceScriptTxOutL .~ maybeToStrictMaybe (fromPlutusScript <$> mkPlutusScript guessTheNumber3)
+
+setupRefTx ::
+  forall era.
+  ( BabbageEraTxOut era
+  , ShelleyEraImp era
+  ) =>
+  ImpTestM era (TxId (EraCrypto era))
+setupRefTx = do
+  shSpending <- setupPlutusV1Script
+  refTxOut <- mkRefTxOut
+  fmap txIdTx . submitTxAnn "Producing transaction" $
+    mkBasicTx mkBasicTxBody
+      & bodyTxL . outputsTxBodyL
+        .~ SSeq.fromList
+          [ refTxOut
+          , scriptLockedTxOut shSpending
+          , scriptLockedTxOut shSpending
+          ]
+
+spec ::
+  forall era.
+  ( ShelleyEraImp era
+  , BabbageEraTxBody era
+  , PredicateFailure (EraRule "LEDGER" era) ~ ConwayLedgerPredFailure era
+  , PredicateFailure (EraRule "UTXOW" era) ~ BabbageUtxowPredFailure era
+  , PredicateFailure (EraRule "UTXO" era) ~ BabbageUtxoPredFailure era
+  ) =>
+  SpecWith (ImpTestState era)
+spec =
+  describe "UTXOS" $ do
+    it "can use reference scripts" $ do
+      producingTx <- setupRefTx
+      referringTx <-
+        submitTxAnn "Transaction that refers to the script" $
+          mkBasicTx mkBasicTxBody
+            & bodyTxL . inputsTxBodyL .~ Set.singleton (mkTxInPartial producingTx 1)
+            & bodyTxL . referenceInputsTxBodyL .~ Set.singleton (mkTxInPartial producingTx 0)
+      (referringTx ^. witsTxL . scriptTxWitsL) `shouldBe` mempty
+    it "can use regular inputs for reference" $ do
+      producingTx <- setupRefTx
+      referringTx <-
+        submitTxAnn "Consuming transaction" $
+          mkBasicTx mkBasicTxBody
+            & bodyTxL . inputsTxBodyL
+              .~ Set.fromList
+                [ mkTxInPartial producingTx 0
+                , mkTxInPartial producingTx 1
+                ]
+      (referringTx ^. witsTxL . scriptTxWitsL) `shouldBe` mempty
+    it "fails with same txIn in regular inputs and reference inputs" $ do
+      producingTx <- setupRefTx
+      let
+        consumingTx =
+          mkBasicTx mkBasicTxBody
+            & bodyTxL . inputsTxBodyL
+              .~ Set.fromList
+                [ mkTxInPartial producingTx 0
+                , mkTxInPartial producingTx 1
+                ]
+            & bodyTxL . referenceInputsTxBodyL .~ Set.singleton (mkTxInPartial producingTx 0)
+      _ <-
+        submitFailingTx
+          consumingTx
+          [ ConwayUtxowFailure . UtxoFailure . BabbageNonDisjointRefInputs @era . NE.singleton $
+              mkTxInPartial producingTx 0
+          ]
+      pure ()
+    it "fails when using inline datums for PlutusV1" $ do
+      shSpending <- setupPlutusV1Script
+      refTxOut <- mkRefTxOut
+      let
```

Comment:
Look at this lonely let! :smile: 
I did not know that his was actually a legal syntax
```suggestion
```

Reply:
Yeah, me neither :smile: